File: rust/src/bin/25.reverse-nodes-in-k-group.rs
/*
 * @lc app=leetcode id=25 lang=rust
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {

        // grab head walk down list reversing the link directions
        // at the kth element we jump back to "head" and link that to the next "head"
        // the kth element was pointing to.
        // repeat

        let mut prev_kth_head: Option<Box<ListNode>> = None;
        let mut kth_head: Box<ListNode> = head.unwrap().as_mut();
        let mut k_index: i32 = 0;
        let mut last: Box<ListNode>;
        let mut new_head: Option<Box<ListNode>> = None;
        let found_head: bool = false;
        let mut node: Option<Box<ListNode>> = head;


        last = kth_head;
        node = last.next;
        k_index += 1;

        while (node.is_some())
        {
            let current = node.unwrap().as_mut();
            k_index += 1;
            if (k_index >= k){
                // if this is the first this node will be the new head of the 
                if (!found_head){
                    new_head = node;
                    found_head = true;
                }
                
                // point prev kth head to current node (will be start of current chain)
                if (prev_kth_head.is_some()){
                    prev_kth_head.unwrap().next = node;
                }
                prev_kth_head = Some(kth_head);

                // link the current kth head to next node - might need to update if chain ends
                k_index = 0;
                kth_head = current;
            }
            else if( k_index == 1){
                last = current;
            }
            else{
                current.next = Some(last);
                last = current;
            }

            node = current.next;
        }

        return new_head;
        
    }
}
// @lc code=end

File: 25.reverse-nodes-in-k-group.py
# @before-stub-for-debug-begin
from python3problem25 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#

# @lc code=start
# Definition for singly-linked list.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # walk through each k sublist of nodes. reverse each connection.
        # keep ref to first node since it will be the "end" of the sublist
        # when we reach the end of a sublist connect the "first" node from before
        # if its the first sublist that node will be the head
        # if we reach the end of the list in the middle of a sublist go back and reverse it

        # trivial case
        if k == 1:
            return head

        newHead: Optional[ListNode]
        index = 0
        current = head
        last: Optional[ListNode] = None
        nextLast = head
        prev = None
        

        while current != None:
            index += 1
            #print(f"index: {index}")
            #print(f"value: {current.val}")
            next = current.next

            if index == k:
                # kth node - last in the sublist
                # first node in the reversed sublist
                # connect the PREVIOUS sublist "last" node
                
                current.next = prev
                if last == None:
                    newHead = current
                else:
                    last.next = current
                last = nextLast
                index = 0
                
                current = next
                continue


            if index == 1:
                # 1st node in sublist
                # will be the last node in the reversed sublist
                nextLast = current
                current.next = None
                # dont set this node next pointer to the prev node
            else:
                # point current node at previous node. Set current node to next node
                current.next = prev
            
            prev = current
            current = next

        if index != 0:
            current = prev
            while current != None:
                next = current.next
                if current == prev:
                    # the last node terminates
                    current.next = None
                else:
                    current.next = prev
                prev = current
                current = next

            # set last sublist end node to the start of this partial sublist.
            last.next = prev

        return newHead

        # debug print modified list
        current = newHead
        while current != None:
            print(f"{current.val}")
            current = current.next

        
        
# @lc code=end

